### PR TITLE
feature - Added table format with a dedicated table-menu

### LINKED ISSUE #17: Add variety of different types of formats
## Description 

The current text editor does not support a variety of different types of formats, which limits its usability. For example, it does not support code blocks, block quotes, or tables. Adding these formats would make the text editor more versatile and useful for a wider range of tasks.

**We can add:**

* CodeBlock
* BlockQuote
* Horizontal Rule
* Tables etc

### PR DESCRIPTION
**Pull request by ashrafchowdury**
## Description

Added **Table format with a dedicated table menu** (closes #17)

The table doesn't show placeholder text for simplicity, and the table menu constantly tracks the table's location by calling a function.

## Added Features
* Table format
* Table menu
* Adding & deleting Column
* Adding & deleting Row
* Delete table

## Preview

https://github.com/steven-tey/novel/assets/87828904/d47a14f7-6032-4187-9f25-b55c8e47e541

### CODE REVIEW

File: ui/editor/index.tsx [hunk 1]
```diff
@@ -140,12 +141,13 @@ export default function Editor() {
       onClick={() => {
         editor?.chain().focus().run();
       }}
-      className="relative min-h-[500px] w-full max-w-screen-lg border-stone-200 p-12 px-8 sm:mb-[calc(20vh)] sm:rounded-lg sm:border sm:px-12 sm:shadow-lg"
+      className=" min-h-[500px] w-full max-w-screen-lg border-stone-200 p-12 px-8 sm:mb-[calc(20vh)] sm:rounded-lg sm:border sm:px-12 sm:shadow-lg"
```
Comment (steven-tey) [inline ui/editor/index.tsx:144]: This breaks the Save Status component's positioning:

CleanShot 2023-06-25 at 16 08 07

**Review by steven-tey — COMMENTED** [review]
Hey @ashrafchowdury! Great work on this PR! Was testing this out and found a few bugs:

1. Clicking on the menu bar doesn't work sometimes. 
2. CMD+A inside a cell doesn't select the contents in the cell.
3. Slash command shouldn't work in a cell. Nor should the `NodeSelector` dropdown.

Here's a screen recording of some of the bugs: 

https://github.com/steven-tey/novel/assets/28986134/9f9ea231-3227-4d26-a733-ca0015063e72

Some potential points of improvement:

1. I'm worried that the current Table Hover menu might not be clear enough for users. Maybe a "hover button to add new columns/rows" approach similar to Notion's might be better: 
CleanShot 2023-06-25 at 16 19 58
2. Ability to resize the columns
3. Ability to rearrange the columns
4. Ability to toggle header column/row on and off: 
CleanShot 2023-06-25 at 16 18 56

**Comment by ashrafchowdury** [comment]
> Hey @ashrafchowdury! Great work on this PR! Was testing this out and found a few bugs:
> 
> 1. Clicking on the menu bar doesn't work sometimes.
> 2. CMD+A inside a cell doesn't select the contents in the cell.
> 3. Slash command shouldn't work in a cell. Nor should the `NodeSelector` dropdown.
> 
> Here's a screen recording of some of the bugs:
> 
>  CleanShot.2023-06-25.at.16.14.06.mp4 
> Some potential points of improvement:
> 
> 1. I'm worried that the current Table Hover menu might not be clear enough for users. Maybe a "hover button to add new columns/rows" approach similar to Notion's might be better:
>    CleanShot 2023-06-25 at 16 19 58
> 2. Ability to resize the columns
> 3. Ability to rearrange the columns
> 4. Ability to toggle header column/row on and off:
>    CleanShot 2023-06-25 at 16 18 56


- Honestly speaking, I'm using Windows 11 and here dose not seen any types of issues which you listed out. I think there is an issue between Mac & windows.


- The hover button to add new columns/rows" approach are bit complex feature, but I can work on it in further future for now I think the menu is perfect.

File: ui/editor/index.tsx (hunk 1 above)
Comment (ashrafchowdury) [reply ui/editor/index.tsx:144]: Ohh, I apologise for that. I'm resolving it now.

**Comment by ashrafchowdury** [comment]
I wanted to add other types of icons for removing columns & rows but lucide icons library doesn't have enough icons.

**Comment by steven-tey** [comment]
> The hover button to add new columns/rows" approach are bit complex feature, but I can work on it in further future for now I think the menu is perfect.

Agreed!

However, these issues are still present and I don't think we can merge this until they're fixed:

1. Once I create a table, there is no way for me to exit it and create a new line.
2. Slash command shouldn't work in a cell. Nor should the `NodeSelector` dropdown.
3. CMD+A inside a cell doesn't select the contents in the cell.

**Comment by mmailaender** [comment]
@ashrafchowdury, are you still working on this? If not, it would be great to get a short feedback, so I can consider handing it over to one of my devs to complete.

**Comment by ashrafchowdury** [comment]
> @ashrafchowdury, are you still working on this? If not, it would be great to get a short feedback, so I can consider handing it over to one of my devs to complete.

I'm not working on it currently, you can freely do what you want.

**Comment by BennyKok** [comment]
This is amazing! Would be great if this can get through, having table support, will make a great fit for the novel-vscode for editing table markdown in vs code.

**Comment by ghoshsanjoy78** [comment]
+1 - Would be awesome to get table support

**Comment by mmailaender** [comment]
I'm expecting to finish this PR in 3 to 4 weeks because we're unfortunately currently in a crunch phase 😉. We have already made the adjustments in our project, but our codebase differs - so we need a calm moment to finish this. If someone can't wait that long, feel free to finish it faster.

**Comment by nepaul** [comment]
Is there a confirmed release date for table support? @mmailaender @steven-tey

**Comment by mmailaender** [comment]
@nepaul, my last post about the timeline is still valid. :) 
Although I need an answer from @steven-tey in [this discussion](https://github.com/steven-tey/novel/discussions/127) about theming support, because our novel version is deeply integrated with theming support. So, depending on @steven-tey decision if he also wants theming support in Novel or not, our next steps will be easier or not.
With theming support: We can create a pull request for table almost as is. (Low effort)
Without theming support: We need to replace our theming support with Novels base theme. (More effort)

**Comment by Gyurmatag** [comment]
Hey, @steven-tey when this table support can get in?

**Pull request closed without merging**